pkg/distributor/forwarding/forwarding.go

```diff
@@ -96,185 +74,260 @@ func NewForwarder(reg prometheus.Registerer, cfg Config) Forwarder {
+	// we need to count these in order to later update some of the distributor's metrics correctly.
+	var notIngestedCounts TimeseriesCounts
+
+	// tsSliceWriteIdx is the index in the toIngest slice where we are writing TimeSeries to.
+	tsSliceWriteIdx := 0
+
+	tsByTargets := f.pools.getTsByTargets()
+	for tsSliceReadIdx, ts := range tsSlice {
+		ingest := false
+		var forwardingTarget string
+
+		metric, err := extract.UnsafeMetricNameFromLabelAdapters(ts.Labels)
+		if err != nil {
+			// Can't check whether a timeseries should be forwarded if it has no metric name.
+			// Send it to the Ingesters and don't forward it.
+			ingest = true
+		} else {
+			rule, ok := rules[metric]
+			if !ok {
+				// There is no forwarding rule for this metric, send it to the Ingesters and don't forward it.
+				ingest = true
+			} else {
+				forwardingTarget = rule.Endpoint
+				ingest = rule.Ingest
+			}
+		}
 
-	rule, ok := r.rules[metric]
-	if !ok {
-		// There is no forwarding rule for this metric, send it to the Ingesters.
-		return true
-	}
-	r.samples.Add(float64(len(sample.Samples)))
+		if forwardingTarget != "" {
+			tsByTarget, ok := tsByTargets[forwardingTarget]
+			if !ok {
+				tsByTarget.ts = f.pools.getTsSlice()
+			}
 
-	ts, ok := r.tsByEndpoint[rule.Endpoint]
-	if !ok {
-		ts = r.pools.timeseries.Get().(*[]mimirpb.PreallocTimeseries)
-		r.requests.Inc()
-	}
+			tsByTarget.ts = f.growTimeseriesSlice(tsByTarget.ts)
+			tsWriteIdx := len(tsByTarget.ts) - 1
+			mimirpb.DeepCopyTimeseries(tsByTarget.ts[tsWriteIdx].TimeSeries, ts.TimeSeries)
+			tsByTarget.counts.count(tsByTarget.ts[tsWriteIdx])
 
-	*ts = append(*ts, sample)
-	r.tsByEndpoint[rule.Endpoint] = ts
+			tsByTargets[forwardingTarget] = tsByTarget
+		}
 
-	return rule.Ingest
-}
+		if ingest {
+			if tsSliceWriteIdx != tsSliceReadIdx {
+				// Swap the timeseries at the reading and writing indices,
+				// later we'll return all timeseries beyond the write index to the pool.
+				tsSlice[tsSliceWriteIdx], tsSlice[tsSliceReadIdx] = tsSlice[tsSliceReadIdx], tsSlice[tsSliceWriteIdx]
+			}
+			tsSliceWriteIdx++
+		} else {
+			notIngestedCounts.count(ts)
+		}
+	}
+
+	// Truncate the toIngest slice to the index up to which we wrote TimeSeries data into it,
+	// all the TimeSeries objects beyond the write index must be returned to the pool.
+	for _, ts := range tsSlice[tsSliceWriteIdx:] {
+		f.pools.putTs(ts.TimeSeries)
+	}
+	tsSlice = tsSlice[:tsSliceWriteIdx]
 
-type recoverableError struct {
-	error
+	return notIngestedCounts, tsSlice, tsByTargets
 }
 
-func (r *request) Send(ctx context.Context) <-chan error {
-	errCh := make(chan error, 1)
+func (f *forwarder) growTimeseriesSlice(ts []mimirpb.PreallocTimeseries) []mimirpb.PreallocTimeseries {
+	newPos := len(ts)
 
-	// Early return if there's no data to send.
-	if len(r.tsByEndpoint) == 0 {
-		close(errCh)
-		return errCh
+	if cap(ts) > len(ts) {
+		ts = ts[:newPos+1]
+	} else {
+		ts = append(ts, mimirpb.PreallocTimeseries{})
 	}
```

Comment:
Can you explain why are we doing this?

Reply:
sorry, now that I look at it again I think it makes no sense, removing it. Thanks for catching.